File: Source/Front-end/front-end/src/app/services/manutenzione/manutenzione.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {environment} from "../../../environments/environment";
import {RichiediManutenzioneRequest} from "../../model/request/richiedi-manutenzione-request";
import {RichiediManutenzioneResponse} from "../../model/response/richiedi-manutenzione-response";
import {ModificaManutenzioneRequest} from "../../model/request/modifica-manutenzione-request";
import {FiltraManutenzioneRequest} from "../../model/request/filtra-manutenzione-request";
import {FiltraManutenzioneResponse} from "../../model/response/filtra-manutenzione-response";
import {AccettaManutenzioneRequest} from "../../model/request/accetta-manutenzione-request";
import {AccettaManutenzioneResponse} from "../../model/response/accetta-manutenzione-response";
import {RiparaManutenzioneRequest} from "../../model/request/ripara-manutenzione-request";
import {RiparaManutenzioneResponse} from "../../model/response/ripara-manutenzione-response";
import {RifiutaManutenzioneRequest} from "../../model/request/rifiuta-manutenzione-request";
import {RifiutaManutenzioneResponse} from "../../model/response/rifiuta-manutenzione-response";
import {FiltroManutenzione} from "../../model/request/filtro-manutenzione";
import {ManutenzioneDTO} from "../../model/response/manutenzione-dto";
import {FeedbackManutenzioneRequest} from "../../model/request/feedback-manutenzione-request";
import {GetManutenzioniDto} from "../../model/response/get-manutenzioni-dto";
import {GetManutenzioneDTO} from "../../model/response/get-manutenzione-dto";

@Injectable({
  providedIn: 'root'
})
export class ManutenzioneService {

  constructor(private http: HttpClient) {
  }

  /**
   * Invia la richiesta di richiedi manutenzione al BE
   */
  richiediManutenzione(richiediManutenzioneRequest: RichiediManutenzioneRequest): Observable<RichiediManutenzioneResponse> {
    return this.http.post<RichiediManutenzioneResponse>(environment.baseUrl + environment.Residente.baseUrl + environment.Manutenzione.richiediManutenzione, richiediManutenzioneRequest);
  }

  /**
   * Invia la richiesta di modifica manutenzione al BE
   */
  modificaManutenzione(modificaManutenzioneRequest: ModificaManutenzioneRequest): Observable<RichiediManutenzioneResponse> {
    return this.http.patch<RichiediManutenzioneResponse>(environment.baseUrl + environment.Residente.baseUrl + environment.Manutenzione.modificaManutenzione, modificaManutenzioneRequest);
  }

  /**
   * Invia la richiesta di filtra manutenzione da parte del supervisore al BE
   */
  filtraManutenzione(filtraManutenzioneRequest: FiltraManutenzioneRequest): Observable<FiltraManutenzioneResponse> {
    return this.http.post<FiltraManutenzioneResponse>(environment.baseUrl + environment.Supervisore.baseUrl + environment.Manutenzione.filtraManutenzione, filtraManutenzioneRequest);
  }

  /**
   * Invia la richiesta di accetta manutenzione al BE
   */
  accettaManutenzione(accettaManutenzioneRequest: AccettaManutenzioneRequest): Observable<AccettaManutenzioneResponse> {
    return this.http.post<AccettaManutenzioneResponse>(environment.baseUrl + environment.Manutenzione.baseUrl + environment.Manutenzione.accettaManutenzione, accettaManutenzioneRequest);
  }

  /**
   * Invia la richiesta di ripara manutenzione al BE
   */
  riparaManutenzione(riparaManutenzioneRequest: RiparaManutenzioneRequest): Observable<RiparaManutenzioneResponse> {
    return this.http.post<RiparaManutenzioneResponse>(environment.baseUrl + environment.Manutenzione.baseUrl + environment.Manutenzione.riparaManutenzione, riparaManutenzioneRequest);
  }

  /**
   * Invia la richiesta di rifiuta manutenzione al BE
   */
  rifiutaManutenzione(rifiutaManutenzioneRequest: RifiutaManutenzioneRequest): Observable<RifiutaManutenzioneResponse> {
    return this.http.post<RifiutaManutenzioneResponse>(environment.baseUrl + environment.Manutenzione.rifiutaManutenzione, rifiutaManutenzioneRequest);
  }

  /**
   * Invia la richiesta di filtro manutenzione al BE
   */
  filtroManutenzioni(filtroManutenzione: FiltroManutenzione): Observable<ManutenzioneDTO[]> {
    return this.http.post<ManutenzioneDTO[]>(environment.baseUrl + environment.All.baseUrl + environment.Manutenzione.filtroManutenzione, filtroManutenzione);
  }

  /**
   * Invia la richiesta di feedback manutenzione al BE
   */
  feedbackManutenzioni(feedbackManutenzioneRequest: FeedbackManutenzioneRequest): Observable<void> {
    return this.http.post<void>(environment.baseUrl + environment.Residente.baseUrl + environment.Manutenzione.feedbackManutenzioni, feedbackManutenzioneRequest);
  }

  /**
   * Invia la richiesta di ritrovamento manutenzioni al BE
   */
  getManutenzioni() {
    return this.http.get<GetManutenzioniDto[]>(environment.baseUrl + environment.All.baseUrl + environment.Manutenzione.getManutenzioni);
  }

  /**
   * Invia la richiesta di ritrovamento manutenzione sulla base dell'id al BE
   */
  getManutenzioneById(id: string) {
    return this.http.get<GetManutenzioneDTO>(environment.baseUrl + environment.All.baseUrl + environment.Manutenzione.getManutenzione + "/" + id);
  }
}

File: Source/Back-end/src/main/java/org/elis/manutenzione/service/definition/ManutenzioneService.java
package org.elis.manutenzione.service.definition;

import org.elis.manutenzione.dto.request.*;
import org.elis.manutenzione.dto.response.GetManutenzioneDTO;
import org.elis.manutenzione.dto.response.GetManutenzioniDTO;
import org.elis.manutenzione.dto.response.ListaManutenzioniResponse;
import org.elis.manutenzione.dto.response.ManutenzioneDTO;
import org.elis.manutenzione.model.entity.Manutentore;
import org.elis.manutenzione.model.entity.Manutenzione;
import org.elis.manutenzione.model.entity.Residente;
import org.elis.manutenzione.model.entity.Utente;

import java.util.List;

/**
 * Interfaccia del servizio per la gestione delle manutenzioni.
 * Questa interfaccia definisce i metodi per le operazioni relative alle manutenzioni,
 * come la richiesta, il filtraggio, l'accettazione, la riparazione, il rifiuto, il feedback e la modifica.
 */
public interface ManutenzioneService {

        /**
         * Richiede una nuova manutenzione.
         *
         * @param richiediManutenzioneRequest DTO contenente i dati della manutenzione da richiedere.
         * @param r                           Il residente che richiede la manutenzione.
         * @return L'oggetto Manutenzione appena creato e salvato.
         */
        public Manutenzione richiediManutenzione(RichiediManutenzioneRequest richiediManutenzioneRequest, Residente r);


        /**
         * Filtra una manutenzione e la assegna ad un manutentore.
         *
         * @param filtraManutenzioneRequest DTO contenente i dati per il filtraggio della manutenzione.
         * @return L'oggetto Manutenzione dopo il filtraggio e l'assegnazione.
         */
        public Manutenzione filtraManutenzione(FiltraManutenzioneRequest filtraManutenzioneRequest);


        /**
         * Segna una manutenzione come riparata.
         *
         * @param riparaManutenzioneRequest DTO contenente i dati per segnare la manutenzione come riparata.
         * @param m                        Il manutentore che ha riparato la manutenzione.
         * @return L'oggetto Manutenzione dopo la riparazione.
         */
        public Manutenzione riparaManutenzione(RiparaManutenzioneRequest riparaManutenzioneRequest, Manutentore m);


        /**
         * Rifiuta una manutenzione.
         *
         * @param rifiutaManutenzioneRequest DTO contenente i dati per il rifiuto della manutenzione.
         * @param u                         L'utente (manutentore o supervisore) che rifiuta la manutenzione.
         * @return L'oggetto Manutenzione dopo il rifiuto.
         */
        public Manutenzione rifiutaManutenzione(RifiutaManutenzioneRequest rifiutaManutenzioneRequest, Utente u);

        /**
         * Accetta una manutenzione.
         *
         * @param accettaManutenzioneRequest DTO contenente i dati per l'accettazione della manutenzione.
         * @param m                        Il manutentore che accetta la manutenzione.
         * @return L'oggetto Manutenzione dopo l'accettazione.
         */
        public Manutenzione accettaManutenzione(AccettaManutenzioneRequest accettaManutenzioneRequest, Manutentore m);


        /**
         * Recupera una lista di manutenzioni filtrate in base ai criteri specificati.
         *
         * @param request DTO contenente i criteri di filtraggio.
         * @param u       L'utente che sta richiedendo la lista di manutenzioni.
         * @return Una lista di oggetti Manutenzione filtrata.
         */
        public List<ManutenzioneDTO> getManutenzioneFiltrata(FiltroManutenzione request, Utente u);

        /**
         * Fornisce un feedback su una manutenzione riparata.
         *
         * @param feedbackManutenzioneRequest DTO contenente i dati del feedback.
         * @param r                           Il residente che fornisce il feedback.
         * @return true se il feedback è stato registrato con successo, false altrimenti.
         */
        public boolean feedbackManutenzione(FeedbackManutenzioneRequest feedbackManutenzioneRequest, Residente r);

        /**
         * Modifica una manutenzione in stato Richiesto.
         *
         * @param request DTO contenente i dati della manutenzione da modificare.
         * @param r       Il residente che ha richiesto la manutenzione.
         * @return L'oggetto Manutenzione modificato.
         */
        public Manutenzione modificaManutenzione(ModificaManutenzioneRequest request, Residente r);

        public List<GetManutenzioniDTO> getManutenzioni(Utente u);

        public GetManutenzioneDTO getManutenzioneById(Long id, Utente u);
}